client/session/manager.go

```diff
@@ -0,0 +1,175 @@
+package session
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+
+	"github.com/pkg/errors"
+	"golang.org/x/net/context"
+	"google.golang.org/grpc"
+)
+
+// Caller can invoke requests on the session
+type Caller interface {
+	Context() context.Context
+	Supports(method string) bool
```

Comment:
It would be nice if we could have both a SupportsMethod / SupportsService method so that service implementers can choose whever to use a per method versioning strategy or a per service one.